### DISCUSSION TITLE
ActivityIndicator not visible bound to IsBusy

### DISCUSSION THREAD

**Discussion by Phil4121** [dotnet/maui#12733] [Q&A]
Hi there,

I am using .Net MAUI in combination with MVVM Toolkit and Community Toolkit focusing on Android right now. At the moment I face the problem that my ActivityIndicator is not visible. 

For the first try I'm using this implementation:

**XAML:**

<ContentPage>

    <ContentPage.Content>

        <Grid RowDefinitions="100,*,Auto">

            <Button
                Grid.Row="2"
                Margin="25,20,25,20"
                Command="{Binding TestCommand}"
                Style="{StaticResource PrimaryButton}"
                Text="TEST" />

            <AbsoluteLayout
                AbsoluteLayout.LayoutBounds="0.5,0.5,1,1"
                AbsoluteLayout.LayoutFlags="All"
                BackgroundColor="#22000000"
                IsVisible="{Binding IsBusy}">

                <ActivityIndicator
                    AbsoluteLayout.LayoutBounds="0.5,0.5,0.1,0.1"
                    AbsoluteLayout.LayoutFlags="All"
                    IsRunning="True"
                    IsVisible="True"
                    Style="{StaticResource activityStackLayout}" />

            </AbsoluteLayout>

        </Grid>

    </ContentPage.Content>

</ContentPage>

**My CS File like this:**

    public partial class AboutPage : ContentPage
    {
        public AboutPage()
        {
            InitializeComponent();
            BindingContext = new AboutViewModel();
        }
    }

**And my ViewModel:**

    public partial class AboutViewModel : BaseViewModel
    {
        public const string ViewName = "AboutPage";
        public AboutViewModel()
        {
            Title = "About";
        }

        [RelayCommand]
        private async Task Test()
        {
            IsBusy = !IsBusy;
        }
    }

In this implementation I can click on the button an my Indicator is visible or not depending on the value of IsBusy. BTW: IsBusy is implemented at BaseViewModel

Now I do a little change in my ViewModel so it looks like this:

    public partial class AboutViewModel : BaseViewModel
    {
        public const string ViewName = "AboutPage";
        public AboutViewModel()
        {
            Title = "About";
        }

        [RelayCommand]
        private async Task Test()
        {
            IsBusy = true;

            var success = await ImportCustomers();

            IsBusy = false;
        }

        Task<bool> ImportCustomers()
        {
            // some heavy operations

            for(int i = 0; i < 10000; i++)
            {
                Console.WriteLine("I: " + i);
            }

            return Task.FromResult(true);
        }
    }

Now my ActivityIndicator isn't visible at all!
I don't think that I've got any typos or a problem with the binding because without calling the async methode everything works fine.

So I started to search around a litte bit and found a few problems with this in the past.
https://github.com/dotnet/maui/issues/6455
https://github.com/dotnet/maui/issues/8617

But all of them seemed to be fixed in the latest version of .Net 7.

Is there still a bug or am I doing something wrong?

Thanks and best regards
Phil

**Comment by mikkpaasuke** [comment]
I am experiencing the same issue with Android (using .NET 7 with VS 2022 Version 17.4.4 and also VS2022 Version 17.5.0 Preview 3.0).

I did notice that if a redraw of the screen is forced (eg if there is an alert displayed via 'Shell.Current.DisplayAlert') within the RelayCommand function then the ActivityIndicator becomes visible, but have yet to progress past this point.

**Reply by Nicobolduc** [reply]
I had the same issue, just don't use the "AbsoluteLayout"

**Reply by arafattehsin** [reply]
Same issue with me too!

**Comment by 3L3F4NT** [comment]
Hey there! I came across this issue and I think I might have a solution that works. If you take a look at the code snippet below, you'll see that all I did was add a couple of await Task.Yield() statements to give the UI a chance to update. After doing that, the ActivityIndicator works perfectly for me. Give it a try and let me know if it works for you too:

```
        [RelayCommand]
        public async Task Login() {

            IsBusy = true;
            await Task.Yield();

            try {
                if (!string.IsNullOrEmpty(UserName) && !string.IsNullOrEmpty(Password)) {

                    // perform login
                }
            }
            catch (Exception ex) {
                // handle ex
            }
            finally {

                IsBusy = false;
                await Task.Yield();
            }
        }
```

Also have a look here: https://github.com/dotnet/maui/issues/12790

**Comment by GrobixDP** [comment]
I had the same issue with the Activity Indicator, spended lots of hours killing my brain cells until i realized that the activity indicator was running on the same thread as my await order, so i created a nuw thread and put the await in that thread. 

await Task.Run(() => viewModel.DownloadDocs(myDocs)); Try this.

**Reply by MauiUIui** [reply]
You're a very good man. Bless you!
